File: evaluation/results/scenario_1b/turing/claude-sonnet-4-20250514/run_03/extracted_code.jl
using Turing, Distributions, CSV, DataFrames, Plots, StatsPlots, Random, LinearAlgebra

# Set random seed for reproducibility
Random.seed!(123)

# Load and prepare the data
function load_data(filename="cases.csv")
    # If the file doesn't exist, create sample data for demonstration
    if !isfile(filename)
        println("Creating sample data for demonstration...")
        dates = Date(2020, 3, 1):Day(1):Date(2020, 6, 30)
        
        # Simulate realistic COVID-19 case trajectory
        n_days = length(dates)
        true_rt = vcat(
            fill(2.5, 20),  # Initial exponential growth
            2.5 .- 0.05 * (1:30),  # Gradual decline
            fill(1.0, 30),  # Around critical threshold
            1.0 .+ 0.02 * (1:20),  # Slight increase
            fill(1.2, n_days - 100)  # Sustained transmission
        )
        
        # Simple simulation for realistic case counts
        cases = zeros(Int, n_days)
        cases[1:7] .= [5, 8, 12, 18, 25, 35, 45]  # Seed cases
        
        for t in 8:n_days
            lambda = sum(cases[max(1, t-14):t-1] .* 
                        exp.(-((1:min(14, t-1)) .- 5.5).^2 / (2 * 2.5^2)))
            expected_cases = max(1.0, true_rt[t] * lambda * 0.1)
            cases[t] = rand(Poisson(expected_cases))
        end
        
        df = DataFrame(date=dates, cases=cases)
        CSV.write(filename, df)
    end
    
    df = CSV.read(filename, DataFrame)
    return df
end

# Define generation interval distribution
function generation_interval_pmf(max_gen=20)
    # Discretized gamma distribution for generation interval
    # Mean ≈ 5.5 days, SD ≈ 2.5 days (typical for COVID-19)
    shape, scale = 4.8, 1.15
    
    pmf = zeros(max_gen)
    for i in 1:max_gen
        pmf[i] = cdf(Gamma(shape, scale), i) - cdf(Gamma(shape, scale), i-1)
    end
    
    # Normalize to ensure sum = 1
    pmf = pmf / sum(pmf)
    return pmf
end

# Define reporting delay distribution
function reporting_delay_pmf(max_delay=15)
    # Log-normal distribution for reporting delay
    # Mean delay ≈ 7 days
    μ, σ = 1.8, 0.5
    
    pmf = zeros(max_delay)
    for i in 1:max_delay
        pmf[i] = cdf(LogNormal(μ, σ), i) - cdf(LogNormal(μ, σ), i-1)
    end
    
    pmf = pmf / sum(pmf)
    return pmf
end

# Turing model for Rt estimation
@model function rt_model(cases, gen_pmf, delay_pmf)
    n_days = length(cases)
    max_gen = length(gen_pmf)
    max_delay = length(delay_pmf)
    
    # Priors for initial infections (seeding period)
    seed_days = 7
    I_seed ~ filldist(Exponential(10.0), seed_days)
    
    # Prior for Rt - using a random walk on log scale for smoothness
    log_rt_init ~ Normal(0.5, 0.5)  # Initial Rt around exp(0.5) ≈ 1.65
    σ_rt ~ Exponential(0.1)  # Innovation standard deviation
    
    log_rt_innovations ~ filldist(Normal(0, 1), n_days - seed_days - 1)
    
    # Construct log_rt time series
    log_rt = Vector{eltype(log_rt_init)}(undef, n_days)
    log_rt[seed_days + 1] = log_rt_init
    
    for t in (seed_days + 2):n_days
        log_rt[t] = log_rt[t-1] + σ_rt * log_rt_innovations[t - seed_days - 1]
    end
    
    # Convert to Rt
    rt = exp.(log_rt[(seed_days + 1):end])
    
    # Initialize infections
    infections = Vector{eltype(I_seed[1])}(undef, n_days)
    infections[1:seed_days] = I_seed
    
    # Renewal equation: compute infections for t > seed_days
    for t in (seed_days + 1):n_days
        lambda = 0.0
        for s in 1:min(max_gen, t-1)
            if t - s >= 1
                lambda += infections[t - s] * gen_pmf[s]
            end
        end
        infections[t] = rt[t - seed_days] * lambda
    end
    
    # Convolve infections with reporting delay to get expected cases
    expected_cases = Vector{eltype(infections[1])}(undef, n_days)
    
    for t in 1:n_days
        expected_cases[t] = 0.0
        for d in 1:min(max_delay, t)
            if t - d + 1 >= 1
                expected_cases[t] += infections[t - d + 1] * delay_pmf[d]
            end
        end
        expected_cases[t] = max(expected_cases[t], 1e-6)  # Avoid zero
    end
    
    # Observation model - Negative Binomial for overdispersion
    φ ~ Exponential(0.1)  # Overdispersion parameter
    
    for t in 1:n_days
        if cases[t] >= 0  # Only observe non-missing cases
            # Negative binomial parameterized by mean and overdispersion
            p = φ / (φ + expected_cases[t])
            r = φ
            cases[t] ~ NegativeBinomial(r, 1 - p)
        end
    end
end

# Function to run the analysis
function estimate_rt(data_file="cases.csv")
    println("Loading data...")
    df = load_data(data_file)
    cases = df.cases
    dates = df.date
    n_days = length(cases)
    
    println("Setting up model...")
    gen_pmf = generation_interval_pmf()
    delay_pmf = reporting_delay_pmf()
    
    println("Generation interval (first 10 days): ", round.(gen_pmf[1:10], digits=3))
    println("Reporting delay (first 10 days): ", round.(delay_pmf[1:10], digits=3))
    
    # Create and sample from the model
    println("Fitting model...")
    model = rt_model(cases, gen_pmf, delay_pmf)
    
    # Use NUTS sampler
    n_samples = 1000
    n_chains = 2
    
    println("Running MCMC sampling...")
    chain = sample(model, NUTS(0.65), MCMCThreads(), n_samples, n_chains)
    
    println("Processing results...")
    
    # Extract Rt estimates
    seed_days = 7
    rt_samples = []
    
    for i in 1:(n_days - seed_days)
        rt_col = "rt[$i]"
        if rt_col in names(chain)
            push!(rt_samples, chain[rt_col])
        end
    end
    
    # Calculate summary statistics
    rt_mean = [mean(samples) for samples in rt_samples]
    rt_q025 = [quantile(vec(Array(samples)), 0.025) for samples in rt_samples]
    rt_q975 = [quantile(vec(Array(samples)), 0.975) for samples in rt_samples]
    
    # Create results dataframe
    rt_dates = dates[(seed_days + 1):end]
    results_df = DataFrame(
        date = rt_dates,
        rt_mean = rt_mean,
        rt_q025 = rt_q025,
        rt_q975 = rt_q975
    )
    
    # Print summary
    println("\nRt Estimation Summary:")
    println("=" ^ 50)
    println("Data period: $(dates[1]) to $(dates[end])")
    println("Total days: $n_days")
    println("Rt estimation period: $(rt_dates[1]) to $(rt_dates[end])")
    println("Mean Rt over period: $(round(mean(rt_mean), digits=2))")
    println("Days with Rt > 1: $(sum(rt_mean .> 1)) / $(length(rt_mean))")
    
    # Create plots
    println("\nCreating plots...")
    
    # Plot 1: Case counts over time
    p1 = plot(dates, cases, 
             title="Daily Case Counts", 
             xlabel="Date", 
             ylabel="Cases",
             linewidth=2,
             color=:blue,
             legend=false)
    
    # Plot 2: Rt over time
    p2 = plot(rt_dates, rt_mean,
             ribbon=(rt_mean .- rt_q025, rt_q975 .- rt_mean),
             title="Estimated Rt Over Time",
             xlabel="Date",
             ylabel="Rt",
             linewidth=2,
             color=:red,
             fillalpha=0.3,
             label="Rt (95% CI)")
    
    hline!([1.0], linestyle=:dash, color=:black, linewidth=2, label="Rt = 1")
    
    # Combined plot
    combined_plot = plot(p1, p2, layout=(2,1), size=(800, 600))
    
    display(combined_plot)
    
    # Save results
    CSV.write("rt_estimates.csv", results_df)
    savefig(combined_plot, "rt_estimation_results.png")
    
    println("\nResults saved to:")
    println("- rt_estimates.csv (Rt estimates)")
    println("- rt_estimation_results.png (plots)")
    
    return results_df, chain, combined_plot
end

# Run the analysis
println("Starting Rt estimation using renewal equation...")
results_df, chain, plot_result = estimate_rt()

# Display final results
println("\nFinal Rt estimates (last 10 days):")
println(last(results_df, 10))

# Model diagnostics
println("\nModel diagnostics:")
println("MCMC Summary:")
display(chain)

File: evaluation/results/scenario_1b/pymc/claude-sonnet-4-20250514/run_01/extracted_code.py
import numpy as np
import pandas as pd
import pymc as pm
import pytensor.tensor as pt
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

# Set style for plots
plt.style.use('seaborn-v0_8')
sns.set_palette("husl")

def load_and_prepare_data(filepath='cases.csv'):
    """Load and prepare the case data"""
    df = pd.read_csv(filepath)
    df['date'] = pd.to_datetime(df['date'])
    df = df.sort_values('date').reset_index(drop=True)
    
    # Remove any negative cases and handle missing values
    df['cases'] = df['cases'].clip(lower=0)
    df = df.dropna()
    
    print(f"Data loaded: {len(df)} days from {df['date'].min()} to {df['date'].max()}")
    print(f"Total cases: {df['cases'].sum():,}")
    print(f"Max daily cases: {df['cases'].max():,}")
    
    return df

def get_generation_interval_pmf(max_days=15, mean_gi=5.2, std_gi=1.72):
    """
    Generate generation interval probability mass function
    Based on COVID-19 literature estimates
    """
    # Use log-normal distribution for generation interval
    days = np.arange(1, max_days + 1)
    
    # Convert mean and std to log-normal parameters
    mu = np.log(mean_gi**2 / np.sqrt(std_gi**2 + mean_gi**2))
    sigma = np.sqrt(np.log(1 + std_gi**2 / mean_gi**2))
    
    # Calculate PMF
    pmf = stats.lognorm.pdf(days, s=sigma, scale=np.exp(mu))
    pmf = pmf / pmf.sum()  # Normalize to sum to 1
    
    print(f"Generation interval: mean={np.sum(days * pmf):.2f}, std={np.sqrt(np.sum(days**2 * pmf) - np.sum(days * pmf)**2):.2f}")
    
    return pmf

def get_reporting_delay_pmf(max_days=21, mean_delay=5.1, std_delay=3.2):
    """
    Generate reporting delay probability mass function
    Time from infection to case reporting
    """
    days = np.arange(0, max_days)
    
    # Use gamma distribution for reporting delay
    # Convert mean and std to shape and scale parameters
    scale = std_delay**2 / mean_delay
    shape = mean_delay / scale
    
    pmf = stats.gamma.pdf(days, a=shape, scale=scale)
    pmf = pmf / pmf.sum()  # Normalize
    
    print(f"Reporting delay: mean={np.sum(days * pmf):.2f}, std={np.sqrt(np.sum(days**2 * pmf) - np.sum(days * pmf)**2):.2f}")
    
    return pmf

def create_rt_model(cases, generation_interval, reporting_delay):
    """
    Create PyMC model for estimating time-varying Rt using renewal equation
    """
    n_days = len(cases)
    max_gi = len(generation_interval)
    max_delay = len(reporting_delay)
    
    with pm.Model() as model:
        # Priors for initial infections (seeding period)
        seed_days = max_gi + max_delay
        log_initial_infections = pm.Normal(
            'log_initial_infections', 
            mu=np.log(np.maximum(cases[:seed_days], 1)), 
            sigma=1.0, 
            shape=seed_days
        )
        initial_infections = pm.Deterministic(
            'initial_infections', 
            pt.exp(log_initial_infections)
        )
        
        # Prior for Rt - assume it varies smoothly over time
        rt_raw = pm.GaussianRandomWalk(
            'rt_raw',
            mu=0,
            sigma=0.1,  # Controls smoothness of Rt over time
            shape=n_days - seed_days
        )
        
        # Transform to positive values with reasonable range
        rt = pm.Deterministic(
            'rt',
            pt.exp(rt_raw + np.log(1.0))  # Centered around 1.0
        )
        
        # Compute infections using renewal equation
        def compute_infections(rt_vals, initial_inf):
            infections = pt.zeros(n_days)
            infections = pt.set_subtensor(infections[:seed_days], initial_inf)
            
            for t in range(seed_days, n_days):
                # Compute convolution sum for renewal equation
                infection_sum = 0
                for s in range(1, min(max_gi + 1, t + 1)):
                    if t - s >= 0:
                        infection_sum += infections[t - s] * generation_interval[s - 1]
                
                infections = pt.set_subtensor(
                    infections[t], 
                    rt_vals[t - seed_days] * infection_sum
                )
            
            return infections
        
        infections = compute_infections(rt, initial_infections)
        
        # Compute expected cases accounting for reporting delay
        def compute_expected_cases(infections_t):
            expected_cases = pt.zeros(n_days)
            
            for t in range(n_days):
                case_sum = 0
                for d in range(min(max_delay, t + 1)):
                    if t - d >= 0:
                        case_sum += infections_t[t - d] * reporting_delay[d]
                expected_cases = pt.set_subtensor(expected_cases[t], case_sum)
            
            return expected_cases
        
        expected_cases = compute_expected_cases(infections)
        
        # Observation model - Negative Binomial to handle overdispersion
        phi = pm.Gamma('phi', alpha=2, beta=0.1)  # Overdispersion parameter
        
        # Convert mean and phi to alpha, beta for NegativeBinomial
        alpha = expected_cases / phi
        
        observed_cases = pm.NegativeBinomial(
            'observed_cases',
            mu=expected_cases,
            alpha=alpha,
            observed=cases
        )
        
        # Store variables for easy access
        model.add_coord('time', values=range(n_days), mutable=True)
        model.add_coord('rt_time', values=range(seed_days, n_days), mutable=True)
    
    return model, seed_days

def fit_model(model, draws=1000, tune=1000, chains=2):
    """Fit the PyMC model"""
    with model:
        # Use NUTS sampler
        trace = pm.sample(
            draws=draws,
            tune=tune,
            chains=chains,
            cores=2,
            target_accept=0.95,
            random_seed=42,
            return_inferencedata=True
        )
    
    return trace

def extract_rt_estimates(trace, dates, seed_days):
    """Extract Rt estimates from the trace"""
    rt_samples = trace.posterior['rt'].values  # shape: (chains, draws, time_points)
    rt_samples_flat = rt_samples.reshape(-1, rt_samples.shape[-1])  # flatten chains and draws
    
    # Calculate summary statistics
    rt_mean = np.mean(rt_samples_flat, axis=0)
    rt_median = np.median(rt_samples_flat, axis=0)
    rt_lower = np.percentile(rt_samples_flat, 2.5, axis=0)
    rt_upper = np.percentile(rt_samples_flat, 97.5, axis=0)
    rt_lower_50 = np.percentile(rt_samples_flat, 25, axis=0)
    rt_upper_50 = np.percentile(rt_samples_flat, 75, axis=0)
    
    # Create results dataframe
    rt_dates = dates[seed_days:]  # Rt estimates start after seeding period
    
    results_df = pd.DataFrame({
        'date': rt_dates,
        'rt_mean': rt_mean,
        'rt_median': rt_median,
        'rt_lower_95': rt_lower,
        'rt_upper_95': rt_upper,
        'rt_lower_50': rt_lower_50,
        'rt_upper_50': rt_upper_50
    })
    
    return results_df

def plot_results(df, rt_results, cases):
    """Plot the results"""
    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(12, 10))
    
    # Plot 1: Cases over time
    ax1.bar(df['date'], df['cases'], alpha=0.7, color='steelblue', label='Observed cases')
    ax1.set_ylabel('Daily cases')
    ax1.set_title('Daily COVID-19 Cases')
    ax1.legend()
    ax1.tick_params(axis='x', rotation=45)
    
    # Plot 2: Rt estimates
    ax2.fill_between(
        rt_results['date'], 
        rt_results['rt_lower_95'], 
        rt_results['rt_upper_95'],
        alpha=0.3, color='red', label='95% CI'
    )
    ax2.fill_between(
        rt_results['date'], 
        rt_results['rt_lower_50'], 
        rt_results['rt_upper_50'],
        alpha=0.5, color='red', label='50% CI'
    )
    ax2.plot(rt_results['date'], rt_results['rt_median'], 'r-', linewidth=2, label='Median Rt')
    ax2.axhline(y=1, color='black', linestyle='--', alpha=0.7, label='Rt = 1')
    ax2.set_ylabel('Reproduction number (Rt)')
    ax2.set_xlabel('Date')
    ax2.set_title('Time-varying Reproduction Number (Rt)')
    ax2.legend()
    ax2.tick_params(axis='x', rotation=45)
    ax2.grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.show()
    
    # Summary statistics
    print("\nRt Summary Statistics:")
    print(f"Mean Rt: {rt_results['rt_mean'].mean():.2f}")
    print(f"Median Rt: {rt_results['rt_median'].median():.2f}")
    print(f"Min Rt (median): {rt_results['rt_median'].min():.2f}")
    print(f"Max Rt (median): {rt_results['rt_median'].max():.2f}")
    
    # Days with Rt > 1
    days_above_1 = (rt_results['rt_median'] > 1).sum()
    total_days = len(rt_results)
    print(f"Days with Rt > 1: {days_above_1}/{total_days} ({100*days_above_1/total_days:.1f}%)")

def main():
    """Main analysis pipeline"""
    print("=== COVID-19 Rt Estimation using Renewal Equation ===\n")
    
    # Load data
    df = load_and_prepare_data('cases.csv')
    cases = df['cases'].values
    
    # Get generation interval and reporting delay distributions
    print("\n" + "="*50)
    print("Setting up epidemiological parameters...")
    generation_interval = get_generation_interval_pmf()
    reporting_delay = get_reporting_delay_pmf()
    
    # Create and fit model
    print("\n" + "="*50)
    print("Creating PyMC model...")
    model, seed_days = create_rt_model(cases, generation_interval, reporting_delay)
    
    print(f"Model created. Seeding period: {seed_days} days")
    print("Model summary:")
    print(model)
    
    print("\n" + "="*50)
    print("Fitting model (this may take several minutes)...")
    trace = fit_model(model, draws=1000, tune=1000, chains=2)
    
    # Extract results
    print("\n" + "="*50)
    print("Extracting Rt estimates...")
    rt_results = extract_rt_estimates(trace, df['date'].values, seed_days)
    
    # Display results
    print("\nFirst 10 Rt estimates:")
    print(rt_results.head(10).round(3))
    
    # Plot results
    print("\n" + "="*50)
    print("Generating plots...")
    plot_results(df, rt_results, cases)
    
    # Save results
    rt_results.to_csv('rt_estimates.csv', index=False)
    print("\nResults saved to 'rt_estimates.csv'")
    
    return rt_results, trace, model

# Example of creating sample data if cases.csv doesn't exist
def create_sample_data():
    """Create sample COVID-19 case data for demonstration"""
    np.random.seed(42)
    dates = pd.date_range('2020-03-01', periods=100, freq='D')
    
    # Simulate epidemic curve with changing Rt
    true_rt = np.concatenate([
        np.linspace(2.5, 0.8, 30),  # Initial decline
        np.linspace(0.8, 1.3, 20),  # Increase
        np.linspace(1.3, 0.7, 30),  # Decline
        np.linspace(0.7, 1.1, 20)   # Slight increase
    ])
    
    # Simple simulation
    cases = []
    infections = [10] * 10  # Initial seeding
    
    gi_pmf = get_generation_interval_pmf()
    
    for t in range(len(dates)):
        if t < 10:
            new_infections = infections[t]
        else:
            infectiousness = sum(infections[max(0, t-s-1)] * gi_pmf[s] 
                               for s in range(min(len(gi_pmf), t)))
            new_infections = np.random.poisson(true_rt[min(t-10, len(true_rt)-1)] * infectiousness)
            infections.append(new_infections)
        
        # Add reporting noise and delay
        reported_cases = max(0, int(new_infections * np.random.uniform(0.7, 1.3)))
        cases.append(reported_cases)
    
    df = pd.DataFrame({
        'date': dates,
        'cases': cases
    })
    
    df.to_csv('cases.csv', index=False)
    print("Sample data created in 'cases.csv'")
    return df

if __name__ == "__main__":
    # Uncomment the next line if you need to create sample data
    # create_sample_data()
    
    try:
        rt_results, trace, model = main()
    except FileNotFoundError:
        print("cases.csv not found. Creating sample data...")
        create_sample_data()
        rt_results, trace, model = main()

